File: python/thirty/thirty.py
def bfs(root, target):
    if root == None:
        return
    queue = [root]
    while queue:
        curr = queue.pop()
        if curr.left != None:
            queue.append(curr.left)
        if curr.right != None:
            queue.append(curr.right)
        queue = queue[1:]
        if curr.val == target:
            return True
        
    return False

def dfs(root, target):
    if root is None:
        return False
    
    stack = [root]
    while stack:
        curr = stack.pop(0)
        if curr.val == target:
            return True
        if curr.left is not None:
            stack.append(curr.left)
        if curr.right is not None:
            stack.append(curr.right)
    return False

def min_height(root):
    if root is None:
        return 0
    
    depth = 1
    queue = [root]

    while queue:
        for _ in range(len(queue)):
            curr = queue.pop(0)
            if curr.left is None and curr.right is None:
                return depth
            if curr.left is not None:
                queue.append(curr.left)
            if curr.right is not None:
                queue.append(curr.right)
        depth += 1
    return depth

class Node():
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def add_two_numbers(l1, l2):
    carry = 0
    dummy = Node()
    helper = dummy
    while l1 or l2 or carry != 0:
        l1_val = l1.val if l1 else 0
        l2_val = l2.val if l2 else 0
        curr_sum = l1_val + l2_val + carry
        val = curr_sum % 10
        carry = carry // 10
        helper.next = Node(val)
        helper = helper.next
        l1 = l1.next if l1 else None
        l2 = l2.next if l2 else None

    return dummy.next

def path_sum(root, target_sum):
    if root is None:
        return False
    stack = [(root, root.val)]

    while stack:
        curr = stack.pop(0)
        if curr[0].left is None and curr[0].right is None:
            if curr[1] == target_sum:
                return True
        else:
            if curr[0].left is not None:
                stack.append((curr[0].left, curr[1] + curr[0].left.val))
            if curr[0].right is not None:
                stack.append((curr[0].right, curr[1] + curr[0].right.val))

    return False

def preorder_traversal(root):
    if root is None:
        return []
    
    values, stack = [], [root]
    while stack:
        curr = stack.pop()
        values.append(curr.val)

        if curr.right is not None:
            stack.append(curr.right)
        if curr.left is not None:
            stack.append(curr.left)
        
    return values

def postorder_traversal(root):
    if root is None:
        return []
    
    values, stack = [], [root]
    while stack:
        curr = stack.pop()
        values.append(curr.val)
        if curr.left is not None:
            stack.append(curr.left)
        if curr.right is not None:
            stack.append(curr.right)
            
    return values[::-1]

def ll_cycle(head):
    if head is None:
        return False
    visited = {}
    curr = head

    while curr:
        if curr in visited:
            return True
        visited[curr] = True
        curr = curr.next

    return False

def ll_intersection(headA, headB):
    if headA is None or headB is None:
        return None
    
    memory = set()
    helperA = headA
    while helperA:
        memory.add(helperA)
        helperA = helperA.next

    helperB = headB
    while helperB:
        if helperB in memory:
            return helperB
        helperB = helperB.next

    return None


def longest_substring(s):
    if s == "":
        return 0
    
    memory = {}
    max_len, lhs = 0, 0

    for rhs in range(len(s)):
        if s[rhs] in memory and memory[s[rhs]] >= lhs:
            lhs = memory[s[rhs]] + 1
        else:
            max_len = max(max_len, rhs - lhs + 1)
        memory[s[rhs]] = rhs
    
    return max_len

File: go/thirty/thirty.go
package thirty

import "slices"

type Node struct {
	Val   int
	Left  *Node
	Right *Node
}

func BFS(root *Node, target int) bool {
	if root == nil {
		return false
	}
	queue := make([]Node, 0)
	queue = append(queue, *root)
	for len(queue) != 0 {
		curr := queue[0]
		if curr.Left != nil {
			queue = append(queue, *curr.Left)
		}
		if curr.Right != nil {
			queue = append(queue, *curr.Right)
		}
		queue = queue[1:]
		if curr.Val == target {
			return true
		}
	}
	return false
}

func MinHeight(root *Node) int {
	if root == nil {
		return 0
	}

	depth := 1
	queue := []*Node{root}

	for len(queue) > 0 {
		queueLen := len(queue)

		for i := 0; i < queueLen; i++ {
			curr := queue[0]
			queue = queue[1:]
			if curr.Left == nil && curr.Right == nil {
				return depth
			}
			if curr.Left != nil {
				queue = append(queue, curr.Left)
			}
			if curr.Right != nil {
				queue = append(queue, curr.Right)
			}
		}
		depth += 1
	}
	return depth
}

type ListNode struct {
	Val  int
	Next *ListNode
}

func AddTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	carry := 0
	dummy := &ListNode{}
	helper := dummy

	for l1 != nil || l2 != nil || carry != 0 {
		var l1Val, l2Val int
		if l1 != nil {
			l1Val = l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			l2Val = l2.Val
			l2 = l2.Next
		}
		sum := l1Val + l2Val + carry
		carry = sum / 10
		helper.Next = &ListNode{sum % 10, nil}
		helper = helper.Next
	}
	return dummy.Next
}

type TreeNode struct {
	Val   int
	Next  *TreeNode
	Left  *TreeNode
	Right *TreeNode
}

type NodeSumPair struct {
	Node *TreeNode
	Sum  int
}

func PathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}
	queue := []*NodeSumPair{{root, root.Val}}

	for len(queue) != 0 {
		curr := queue[0]
		queue = queue[1:]
		if curr.Node.Left != nil && curr.Node.Right != nil {
			if curr.Sum == targetSum {
				return true
			}
		} else {
			if curr.Node.Left != nil {
				queue = append(queue, &NodeSumPair{curr.Node.Left, curr.Sum + curr.Node.Left.Val})
			}
			if curr.Node.Right != nil {
				queue = append(queue, &NodeSumPair{curr.Node.Right, curr.Sum + curr.Node.Right.Val})
			}
		}
	}
	return false
}

func PreorderTraversal(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	var values []int
	stack := []TreeNode{*root}

	for len(stack) > 0 {
		curr := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		values = append(values, curr.Val)
		if curr.Right != nil {
			stack = append(stack, *curr.Right)
		}
		if curr.Left != nil {
			stack = append(stack, *curr.Left)
		}
	}

	return values
}

func PostorderTraversal(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	var values []int
	stack := []TreeNode{*root}

	for len(stack) > 0 {
		curr := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		values = append(values, curr.Val)
		if curr.Left != nil {
			stack = append(stack, *curr.Left)
		}
		if curr.Right != nil {
			stack = append(stack, *curr.Right)
		}
	}
	slices.Reverse(values)
	return values
}

func LLCycle(head *ListNode) bool {
	visited := make(map[*ListNode]bool)
	curr := head

	for curr != nil {
		if visited[curr] {
			return true
		}
		visited[curr] = true
		curr = curr.Next
	}
	return false
}

func LLIntersection(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}

	memory := make(map[*ListNode]bool)
	helperA := headA
	for helperA != nil {
		memory[helperA] = true
	}

	helperB := headB
	for helperB != nil {
		if _, exists := memory[helperB]; exists {
			return helperB
		}
		helperB = headB.Next
	}
	return nil
}

func LongestSubstring(s string) int {
	if len(s) == 0 {
		return 0
	}
	memory := make(map[string]int, 0)
	lhs := 0
	result := 0

	for rhs := 0; rhs < len(s); rhs++ {
		if _, exists := memory[string(s[rhs])]; exists && memory[string(s[rhs])] >= lhs {
			lhs = memory[string(s[rhs])] + 1
		} else {
			result = max(result, rhs-lhs+1)
		}
		memory[string(s[rhs])] = rhs
	}
	return result
}

func MaxArea(heights []int) int {
	lhs := 0
	rhs := len(heights)
	if rhs == 0 {
		return 0
	}
	rhs -= 1

	maxArea := 0

	for lhs != rhs {
		base := rhs - lhs
		h := min(heights[lhs], heights[rhs])
		maxArea = max(maxArea, base*h)
		if heights[lhs] <= heights[rhs] {
			lhs += 1
		} else if heights[rhs] <= heights[lhs] {
			rhs -= 1
		} else {
			lhs += 1
		}
	}
	return maxArea
}
